apis/v1alpha1/deploykey_types.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+)
+
+// DeployKeyParameters define the desired state of a Gitlab deploy key
+type DeployKeyParameters struct {
+	// ProjectID is the ID of the project to create the deploy key in.
+	// +optional
+	// +immutable
+	ProjectID *int `json:"projectId,omitempty"`
+
+	// ProjectIDRef is a reference to a project to retrieve its projectId
+	// +optional
+	// +immutable
+	ProjectIDRef *xpv1.Reference `json:"projectIdRef,omitempty"`
+
+	// ProjectIDSelector selects reference to a project to retrieve its projectId.
+	// +optional
+	ProjectIDSelector *xpv1.Selector `json:"projectIdSelector,omitempty"`
+
+	// Title is the new deploy key’s title
+	Title *string `json:"title,omitempty"`
+
+	// Key is the new public deploy key. If left empty, a random key will be generated
+	// and the private part written to writeConnectionSecretToRef
+	// +optional
+	Key *string `json:"key,omitempty"`
+
+	// CanPush defines whether the deploy key can push to the project’s repository
+	// +optional
+	CanPush *bool `json:"canPush,omitempty"`
+}
+
+// DeployKeyObservation represents a deploy key.
+//
+// GitLab API docs:
+// https://docs.gitlab.com/ee/api/deploy_keys.html
+type DeployKeyObservation struct {
+	// ID of the deploy key at gitlab
+	ID int `json:"id,omitempty"`
+
+	// CreatedAt specifies the time the deploy key was created
+	CreatedAt *metav1.Time `json:"createdAt,omitempty"`
+}
+
+// A DeployKeySpec defines the desired state of a Gitlab DeployKey.
+type DeployKeySpec struct {
+	xpv1.ResourceSpec `json:",inline"`
+	ForProvider       DeployKeyParameters `json:"forProvider"`
+}
+
+// A DeployKeyStatus represents the observed state of a Gitlab DeployKey.
+type DeployKeyStatus struct {
+	xpv1.ResourceStatus `json:",inline"`
+	AtProvider          DeployKeyObservation `json:"atProvider,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// A DeployKey is a managed resource that represents a Gitlab deploy key
+// +kubebuilder:printcolumn:name="READY",type="string",JSONPath=".status.conditions[?(@.type=='Ready')].status"
+// +kubebuilder:printcolumn:name="SYNCED",type="string",JSONPath=".status.conditions[?(@.type=='Synced')].status"
+// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
```

Comment:
I would use `externalName` instead of `creationTimestamp` as this is much more meaningful information for the user.

```golang
// +kubebuilder:printcolumn:name="EXTERNAL-NAME",type="string",JSONPath=".metadata.annotations.crossplane\\.io/external-name"
```